hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java

```diff
@@ -61,6 +69,63 @@ public void remove() {
         };
     }
 
+    /**
+     * Lazily filters iterated elements.
+     *
+     * @return a new iterable object which
+     * has an iterator capable of filtering elements
+     */
+    public static <T> Iterable<T> filter(Iterable<T> iterable, Predicate<T> filter) {
+        Iterator<T> givenIterator = iterable.iterator();
+
+        Iterator<T> filteringIterator = new Iterator<T>() {
+            private T next;
+
+            @Override
+            public boolean hasNext() {
+                boolean hasNext = false;
+                while (givenIterator.hasNext()) {
```

Comment:
If I call `hasNext()` multiple times it shouldn't advance the iterator, but it seems it does.

Reply:
good catch, fixed.